lib/srv/db/spanner/engine.go

```diff
@@ -0,0 +1,169 @@
+	sessionCtx *common.Session
+	// instanceID is used for authz and is of the form
+	// projects/<project-id>/instances/<instance-id>/,
+	// where the project ID and instance ID come from the Teleport database's
+	// GCP metadata.
+	// Every RPC that specifies a database name must be prefixed by this or the
+	// RPC will be denied.
+	instanceID string
+
+	// mu guards access to gcloudClient
+	mu sync.Mutex
+	// gcloudConn is an upstream client to Google cloud API used to forward RPCs
+	// to the real Spanner API.
+	gcloudClient gtransport.ConnPool
+
+	// waitForSessionStart is closed after the session start event has been
+	// emitted.
+	waitForSessionStart chan struct{}
+	// sessionStartErr is the error (or nil) from the first RPC access check
+	// that starts the session.
+	sessionStartErr error
+	// startSessionOnce is used to start the session only once.
+	startSessionOnce sync.Once
+
+	// UnimplementedSpannerServer is embedded for forward interface compat.
+	spannerpb.UnimplementedSpannerServer
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	gcp := sessionCtx.Database.GetGCP()
+	e.instanceID = "projects/" + gcp.ProjectID + "/instances/" + gcp.InstanceID + "/"
+	e.sessionCtx = sessionCtx
+	e.waitForSessionStart = make(chan struct{})
+	return nil
+}
+
+func (e *Engine) SendError(err error) {
+	if err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+	// the grpc server handles sending all errors, if an error is sent outside
+	// of that, just log it here.
+	e.Log.WithError(err).Debug("GCP Spanner connection error")
+}
+
+// HandleConnection processes the connection from the proxy coming over reverse
+// tunnel.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	observe := common.GetConnectionSetupTimeObserver(e.sessionCtx.Database)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	e.clientConnContext = ctx
+	e.cancelContext = cancel
+
+	// stand up a gRPC grpcServer to handle the client connection.
+	grpcServer := grpc.NewServer(
+		grpc.ChainUnaryInterceptor(e.unaryServerInterceptors()...),
+		grpc.ChainStreamInterceptor(e.streamServerInterceptors()...),
+		grpc.MaxConcurrentStreams(defaults.GRPCMaxConcurrentStreams),
+	)
+	// register server services to proxy RPCs.
+	spannerpb.RegisterSpannerServer(grpcServer, e)
+
+	// this doesn't block, because the listener returns when Accept is called
+	// for a second time.
+	err := grpcServer.Serve(newSingleUseListener(e.clientConn, observe))
+	if err != nil && !errors.Is(err, io.EOF) {
+		return trace.Wrap(err)
+	}
+	select {
+	case <-ctx.Done():
+	case <-e.Context.Done():
+	}
+
+	// this will cause the server to reject all new RPCs and block until all
+	// outstanding handlers have returned.
+	grpcServer.GracefulStop()
+	if e.gcloudClient != nil {
+		e.gcloudClient.Close()
+	}
```

Comment:
If the monitoring context is closed say because of a lock or expired cert, will the graceful stop wait until the streaming apis finish?

Reply:
It will wait for all handlers to return - the monitor will close the connection, so the streaming should be interrupted by that and the handler should return.

I can also add a select in `proxyServerStream` though, just to be sure

Reply:
yes. Like you said, since the monitor will close the con then this should be fine 👍